tools/py_common/binary_data.py

```diff
@@ -0,0 +1,274 @@
+
+    marker = b[0]
+    if marker < NEG_MULTI_MARKER or marker >= 0xf0:
+        raise ValueError('Not a packed integer')
+    elif marker < NEG_2BYTE_MARKER:
+        sz = 8 - get_bits(marker, 4)
+        if sz < 0:
+            raise ValueError('Not a valid packed integer')
+        part1 = (-1 << (sz << 3))
+        part2 = get_int(b[1:], sz)
+        part3 = b[sz+1:]
+        return (part1 | part2, part3)
+    elif marker < NEG_1BYTE_MARKER:
+        return (NEG_2BYTE_MIN + ((get_bits(marker, 5) << 8) | b[1]), b[2:])
+    elif marker < POS_1BYTE_MARKER:
+        return (NEG_1BYTE_MIN + get_bits(marker, 6), b[1:])
+    elif marker < POS_2BYTE_MARKER:
+        return (get_bits(marker, 6), b[1:])
+    elif marker < POS_MULTI_MARKER:
+        return (POS_1BYTE_MAX + 1 +
+               ((get_bits(marker, 5) << 8) | b[1]), b[2:])
+    else:
+        sz = get_bits(marker, 4)
+        return (POS_2BYTE_MAX + 1 + get_int(b[1:], sz), b[sz+1:])
+
+
+def decode_esc_hex(s: str) -> bytes:
+    '''
+    Decode a string produced by __wt_raw_to_esc_hex.
+    '''
+    b = bytearray()
+    i = 0
+    while i < len(s):
+        if s[i] == '\\':
+            if i + 3 > len(s):
+                raise ValueError('Not a valid escaped hex byte')
+            b.append(int(s[i + 1] + s[i + 2], 16))
+            i += 3
+        else:
+            b.append(ord(s[i]))
+            i += 1
+    return bytes(b)
+
+
+class FileAsArray:
+    '''
+    A wrapper class around a file-like object that allows it to be treated as an array.
+    '''
+
+    def __init__(self, fileobj):
+        '''
+        Initialize the class.
+        '''
+        self.fileobj = fileobj
+        self.pos = 0
+        self.slicepos = 0
+
+    def __getitem__(self, key: typing.Union[int, slice]) -> bytes:
+        '''
+        Get the item based on the given key: an integer that specifies the offset, or a slice to
+        read a range of bytes.
+        '''
+        if isinstance(key, slice):
+            # We only handle limited slices, like this: x[n:]
+            if key.stop != None and key.step != None:
+                raise ValueError('slice error: ' + str(key))
+            ret = FileAsArray(self.fileobj)
+            ret.slicepos = key.start
+            self.pos = -1
+            return ret
+        elif isinstance(key, int):
+            if self.pos != key:
+                self.fileobj.seek(key + self.slicepos)
+                self.pos = key
+            ret = self.fileobj.read(1)[0]
+            self.pos += 1
+            return ret
+        else:
+            raise ValueError('not implemented: ' + str(key))
+
+
+class BinaryFile(object):
+    '''
+    An encapsulation around a file object that saves bytes read in increments so raw bytes can be
+    shown.
+    '''
+
+    def __init__(self, fileobj):
+        '''
+        Initialize the instance of this binary stream.
+        '''
+        self.fileobj = fileobj
+        self.saved = bytearray()
+
+
```

Comment:
Isn't a single blank line between functions/methods recommended?  (Maybe double blank between classes?). This file uses both 1 line and 2 lines between functions/methods.

Reply:
Good point—I'll adjust this to what the style guide says:

> Surround top-level function and class definitions with two blank lines.
> 
> Method definitions inside a class are surrounded by a single blank line.

Source: https://peps.python.org/pep-0008/